SEMANTICS.md

```diff
@@ -294,6 +294,9 @@ returns `data: None`, it will not modify any previously set data state. If there
 are multiple submessages all setting this, only the last one is used (they all
 overwrite any previous `data` value).
 
+An edge is *empty* data returned with `data: Some(empty)`. We handle this equal
+to `data: None` to avoid unmarshalling issues.
```

Comment:
With this solution, there is no way for a contract to clear previous data. E.g.

```
// Execute
data: Some(b"first thought")

// Reply
// Okay, better shut my mouth
data: Some(b"")
```

Do we really want to do that? Or better increase testing for the null/empty differentiation?

Reply:
I cannot conceive of such a case, but I do admit there are cases we thought would never come up, which later do.

That said, the whole `Some(b"")` vs `None` his us in many places, and I am still not convinced that in any of them we really needed to differentiate them.

Where we pass these values in JSON, we need to explicitly use custom types in Go, as the default parser for `[]byte` doesn't distinguish between these two "zero values".

Reply:
> Where we pass these values in JSON, we need to explicitly use custom types in Go, as the default parser for `[]byte` doesn't distinguish between these two "zero values".

I think we tested that and it worked well. It just collapses with strings as the `string` type in Go is not nullable but `[]byte` is.

Reply:
Hmmm... you are right... This does produce different values for `null`/missing and `""`
 https://play.golang.org/p/KqI86DXQ7jq

I guess it makes sense to test this and enforce different behaviors.

I was concerned about distinguishing them and advised Alex to be careful about handling these two cases differently